src/main/java/cam72cam/immersiverailroading/util/CAML.java

```diff
@@ -0,0 +1,172 @@
+package cam72cam.immersiverailroading.util;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.commons.lang3.StringUtils;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+import java.util.*;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+/*
+ * CAML =
+ *      # Cam's Awesome Markup Language V1
+ *      description: CAML is an incredibly simple markup language, with a very small rule set.
+ *      description: It is an indentation based format, which allows both tabs and spaces, as long as they are consistent.
+ *      formatting =
+ *          comment = "Any character that follows a # is ignored, unless it is a contained in a quoted string"
+ *          whitespace = Any combination of spaces and tabs
+ *          key =
+ *              # An identifier made up of any non-whitespace character followed by a '=' or ':'
+ *              set: When a key is followed by ':' there can be multiple keys with that name which are grouped in a collection
+ *              single = When a key is followed by '=' there can only be a single key with that name.
+ *          string = Any sequence of characters that is optionally wrapped in "quotation marks"
+ *      types =
+ *          # There are two types in CAML
+ *          blocks =
+ *              description = A block is a named container
+ *              format = A block is started with a key at the current indentation level
+ *              contents = Any subsequent lines at a deeper indentation level are read and then added to the block as their respective types.
+ *
+ *          primitive = A key followed by a value
+ *
+ * */
+public class CAML {
+    private static final Pattern base = Pattern.compile("(\\s*)(\\S+)\\s*([=:])\\s?(.*)");
+    public static DataBlock parse(InputStream stream) throws IOException {
+        List<String> lines = IOUtils.readLines(stream, StandardCharsets.UTF_8).stream()
+                .map(s -> s.replaceFirst("#.*", ""))
+                .filter(s -> !StringUtils.isWhitespace(s))
+                .collect(Collectors.toList());
+
+        return new Block(lines);
+    }
+
+    private static class Block implements DataBlock {
+        private final Map<String, Value> primitives = new LinkedHashMap<>();
+        private final Map<String, List<Value>> primitiveSets = new LinkedHashMap<>();
+        private final Map<String, DataBlock> blocks = new LinkedHashMap<>();
+        private final Map<String, List<DataBlock>> blockSets = new LinkedHashMap<>();
+
+        private Block(List<String> lines) throws IOException {
```

Comment:
These propbably shouldn't be IOExceptions, I think there's a Std Lib exception for parsing errors.

Reply:
Fair point